File: ProjectEuler/p30.cc
//author aghis daniels
//Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.
#include <iostream> //cout

using namespace std;

//int rec_sum(int n)
//recursive version of the function below
//x is the number we will break up and sum
//return the sum
int rec_sum(int n){
	if(1>=n){
		return n*n*n*n*n;
	}else{
		return (n%10)*(n%10)*(n%10)*(n%10)*(n%10)+rec_sum(n/10);
	}
}

//int sum_of_num(int n)
//a traditional number spliting functions powers of 5ths
//x is the number we will break up and sum
//return the sum
int sum_of_num(int n){
	int sum=0;
	while(1<n){
		const int tmp=n%10;
		n/=10;
		sum+=tmp*tmp*tmp*tmp*tmp;
	}
	sum+=n*n*n*n*n;
	return sum;
}

//bool is_num(const int in)
//a helper function
//x is the number we will break up and sum
//return true is in == sum of digit fifths
bool is_num(const int in){
	return (in==sum_of_num(in));
}

//int solve()
//solving the problem here
int solve(){
	int ans=0;
	for(int i=2; i<250000; ++i){
		if(is_num(i)){
			ans+=i;
		}
	}
	return ans;
}

int main(){
    int total=solve();
    cout<<"the sum of all numbers taht digits are sum of powers of 5 are "<<total<<endl;
    return 0;
}

File: ProjectEuler/Python/p30.py
#author aghis daniels
#Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.

#digits(x)
#a compact but slower version of summing the powers of fifths
#x is the number we will break up and sum
#return the sum
def digits(x):
	return sum([int(i)**5 for i in str(x)])

#digits2(x)
#a traditional number spliting functions powers of 5ths
#x is the number we will break up and sum
#return the sum
def digits2(x):
	rsum=0
	while 1<x:
		tmp=x%10
		x//=10
		rsum+=tmp*tmp*tmp*tmp*tmp
	rsum+=x*x*x*x*x
	return rsum

#is_num(x)
#helper for comparing x and sum of fifths
#x is the number we will break up and sum
#return true if x = sum of fifths
def is_num(x):
	return(x==digits2(x))

#solve()
#this will run throuigh all the possible solutions
#param none
#return answers 	
def solve():
	return sum([i for i in range(2,200000)if is_num(i)])
	
def main():
	print(solve())

if __name__=='__main__':
    main()
